File: leetcode/problems/problem-specific/746.min-cost-climbing-stairs/746.min-cost-climbing-stairs.go
package main

/*
 * @lc app=leetcode id=746 lang=golang
 *
 * [746] Min Cost Climbing Stairs
 */

// @lc code=start

// * Recursive Memoization Solution | O(n) Time | O(n) Space
// * n -> Length of the cost array

func minCostClimbingStairs(cost []int) int {
	return findMin(minCostClimbingStairsHelper(cost, make([]int, len(cost)+1), 0),
		minCostClimbingStairsHelper(cost, make([]int, len(cost)+1), 1))
}

func minCostClimbingStairsHelper(cost, memcache []int, curStep int) int {
	if curStep >= len(cost) {
		return 0
	}
	if memcache[curStep] > 0 {
		return memcache[curStep]
	}

	memcache[curStep] = cost[curStep] + findMin(minCostClimbingStairsHelper(cost, memcache, curStep+1),
		minCostClimbingStairsHelper(cost, memcache, curStep+2))
	return memcache[curStep]
}

func findMin(a, b int) int {
	if a <= b {
		return a
	}

	return b
}

// @lc code=end

File: leetcode/problems/problem-specific/746.min-cost-climbing-stairs/746.min-cost-climbing-stairs.py
#
# @lc app=leetcode id=746 lang=python3
#
# [746] Min Cost Climbing Stairs
#

# @lc code=start

# * Iterative Bottom-Up Solution | O(n) Time | O(1) Space
# * n -> The length of cost array


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        for i in range(2, len(cost)):
            cost[i] += min(cost[i - 1], cost[i - 2])

        return min(cost[-1], cost[-2])


# @lc code=end
